src/orange/orangeinp/IntersectRegion.cc

```diff
@@ -334,6 +334,233 @@ void Ellipsoid::output(JsonPimpl* j) const
+                               real_type halfheight)
+    : lower_radii_{lower_radii}, upper_radii_{upper_radii}, hh_{halfheight}
+{
+    SoftZero soft_zero;
+    SoftEqual soft_equal;
+
+    auto X = to_int(Axis::x);
+    auto Y = to_int(Axis::y);
+
+    // True if either radius is negative
+    auto has_negative
+        = [&](Real2 const& radii) { return radii[X] < 0 || radii[Y] < 0; };
+
+    // True if radii is (0, 0)
+    auto is_vertex = [&](Real2 const& radii) {
+        return soft_zero(radii[X]) && soft_zero(radii[Y]);
+    };
+
+    // True if radii is (0, x) || (x, 0), where x != 0
+    auto is_partial_zero = [&](Real2 const& radii) {
+        return (soft_zero(radii[X]) != soft_zero(radii[Y]));
+    };
+
+    // Check for negatives
+    CELER_VALIDATE(!has_negative(lower_radii_),
+                   << "Lower radii cannot have negative values: "
+                   << lower_radii_[X] << ", " << lower_radii_[Y]);
+    CELER_VALIDATE(!has_negative(upper_radii_),
+                   << "Upper radii cannot have negative values: "
+                   << upper_radii_[X] << ", " << upper_radii_[Y]);
+
+    // Check for partial zeros
+    CELER_VALIDATE(!is_partial_zero(lower_radii_),
+                   << "Mismatched zero lower radii: " << lower_radii_[X]
+                   << ", " << lower_radii_[Y]);
+    CELER_VALIDATE(!is_partial_zero(upper_radii_),
+                   << "Mismatched zero upper radii: " << upper_radii_[X]
+                   << ", " << upper_radii_[Y]);
+
+    // Check aspect ratios
+    if (!is_vertex(lower_radii_) && !is_vertex(upper_radii_))
+    {
+        CELER_VALIDATE(soft_equal(lower_radii_[X] / lower_radii_[Y],
+                                  upper_radii_[X] / upper_radii_[Y]),
+                       << "Upper and lower elliptical cross sections must "
+                          "have the same aspect ratio.");
+    }
+
+    // Check for elliptical cylinders. Since we have already validated the
+    // aspect ratio, we only need to test the x-values here.
+    CELER_VALIDATE(!soft_equal(lower_radii_[X], upper_radii_[X]),
+                   << "Upper and lower elliptical cross sections must "
+                      "not be equal");
+
+    // Check positivity of half-height
+    CELER_VALIDATE(hh_ > 0, << "Nonpositive halfheight: " << hh_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Whether this encloses another elliptical cone.
+ */
+bool EllipticalCone::encloses(EllipticalCone const& other) const
+{
+    for (auto ax : range(2))
+    {
+        if (this->lower_radii_[ax] < other.lower_radii_[ax]
+            || this->upper_radii_[ax] < other.upper_radii_[ax])
+        {
+            return false;
+        }
+    }
+    return hh_ >= other.hh_;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build surfaces.
+ *
+ * The elliptical surface can be expressed as [1]:
+ *
+ * (x/r_x)^2 + (y/r_y)^2 = (v-z)^2,
```

Comment:
Use doxygen `\f[` markers here and probably a `\verbatim` block below. Same comment about the elliptical cone source (and use doxygen markups for that too.

Reply:
Both of your comments are on elliptical cone. I have moved the all the info from the .cc file in the class-level doc string.